openglyph/src/io/mega_file.cpp

```diff
@@ -0,0 +1,170 @@
+#include <khepri/math/math.hpp>
+#include <khepri/utility/crc.hpp>
+#include <khepri/utility/string.hpp>
+
+#include <openglyph/io/mega_file.hpp>
+
+#include <iostream>
+namespace openglyph::io {
+
+namespace {
+void verify(bool condition)
+{
+    if (!condition) {
+        throw khepri::io::InvalidFormatError();
+    }
+}
+} // namespace
+
+/**
+ * @brief this class represents a subfile inside the megafile archive.
+ *
+ * @note This instance may not outlive its parent megafile archive.
+ */
+class MegaFile::SubFile : public khepri::io::Stream
+{
+public:
+    SubFile(const SubFileInfo& info, khepri::io::File* file);
+
+    SubFile(const SubFile&)            = delete;
+    SubFile(SubFile&&)                 = delete;
+    SubFile& operator=(const SubFile&) = delete;
+    SubFile& operator=(SubFile&&)      = delete;
+
+    /// \see stream::readable
+    [[nodiscard]] bool readable() const noexcept override
+    {
+        return true;
+    }
+
+    /// \see stream::writable
+    [[nodiscard]] bool writable() const noexcept override
+    {
+        return false;
+    }
+
+    /// \see stream::seekable
+    [[nodiscard]] bool seekable() const noexcept override
+    {
+        return true;
+    }
+
+    /// \see stream::read
+    virtual size_t read(void* buffer, size_t count) override;
+
+    /// \see stream::write
+    virtual size_t write(const void* buffer, size_t count) override;
+
+    /// \see stream::seek
+    virtual long long seek(long long offset, khepri::io::SeekOrigin origin) override;
+
+private:
+    const MegaFile::SubFileInfo info;
+    khepri::io::File*           m_mega_file;
+    std::uint64_t               m_local_read_offset;
+};
+
+MegaFile::MegaFile(const std::filesystem::path& mega_file_path)
+    : m_file(std::make_unique<khepri::io::File>(mega_file_path, khepri::io::OpenMode::read))
+{
+    std::tie(m_filenames, m_fileinfo) = extract_metadata();
+}
+
+std::unique_ptr<khepri::io::Stream> MegaFile::open_file(const std::filesystem::path& path)
+{
+    std::uint32_t crc            = khepri::CRC32::calculate(khepri::uppercase(path.string()));
+    std::string   uppercase_path = khepri::uppercase(path.string());
```

Comment:
Nit: can make these two `const`. Also, you can reuse `uppercase_path` for `crc`.